dashboard/test/models/course_offering_test.rb

```diff
@@ -213,6 +213,24 @@ class CourseOfferingTest < ActiveSupport::TestCase
     assert course_offering.valid?
   end
 
+  test "latest_published_version returns most recent published course version" do
+    offering = create :course_offering
+
+    most_recent_version = create :course_version, course_offering: offering, key: '2021'
+    most_recent_version.content_root.update!(published_state: 'beta')
+
+    preview_version = create :course_version, course_offering: offering, key: '2019'
+    preview_version.content_root.update!(published_state: 'preview')
+
+    most_recent_published_version = create :course_version, course_offering: offering, key: '2020'
+    most_recent_published_version.content_root.update!(published_state: 'preview')
+
+    stable_version = create :course_version, course_offering: offering, key: '2018'
+    stable_version.content_root.update!(published_state: 'stable')
+
+    assert_equal offering.latest_published_version, most_recent_published_version
+  end
```

Comment:
nice test! I'm glad to see the factory is working, though it is still sad that you have to know that course version is the factory to use if you want it to create the other stuff for you.

I think it would be helpful to add an similar test where all of the content roots are `with_unit`, and update the name of this test to make it clear that it is hitting the unit group case, since the default factory behavior is not obvious here.

Reply:
Yeah I hear that –– sounds good, will do in a follow-up PR!

Reply:
Oh wait for some reason I thought I had merged this –– I'll do this in this PR!

Reply:
Added a test to address `unit`. For getting the course version factory to do what we're wanting, I had tried `association :course_version` before I realized that the association goes the other way. I think I'd need to do something like `course_version {create :course_version}` –– I can think about this in the follow-up validation PR.

Reply:
I think I am missing a piece here. which factory did you mean adding `association :course_version` and `course_version {create :course_version}` to? and/or did you mean `course_offering` in any of the places it says `course_version`?

Reply:
I think I meant `course_version` for all of these. A course offering `has many` course versions. Inside the course version factory, we do `association :course_offering`.

I had originally tried to do `association :course_version` inside the course offering factory, but the association goes the other way. But for a course offering to correctly run `course_offering.course_versions` I'm wondering if we need to add course_versions explicitly, like `course_version {create :course_version}`, similar to this:
https://github.com/code-dot-org/code-dot-org/blob/6519c6c92a0fc29f783d7ef5cfd3dd65c9a998f6/dashboard/test/factories/pd_factories.rb#L517

That's for one, but for the has_many association, it looks similar but with a loop, e.g.
https://thoughtbot.com/blog/aint-no-calla-back-bot

Reply:
your goal is to do something like `create :course_offering, :with_units` right? 

> I had originally tried to do association :course_version inside the course offering factory, but the association goes the other way.

I'm not sure I understand all of the paths you have explored, but this works for me:
```diff
diff --git a/dashboard/test/factories/factories.rb b/dashboard/test/factories/factories.rb
index dd3e14249a7..40c9ee0ecf2 100644
--- a/dashboard/test/factories/factories.rb
+++ b/dashboard/test/factories/factories.rb
@@ -5,6 +5,13 @@ FactoryBot.define do
     sequence(:key, 'a') {|c| "bogus-course-offering-#{c}"}
     sequence(:display_name, 'a') {|c| "bogus-course-offering-#{c}"}
     assignable {true}
+
+    trait :with_units do
+      after(:create) do |course_offering|
+        create(:course_version, :with_unit, course_offering: course_offering)
+        create(:course_version, :with_unit, course_offering: course_offering)
+      end
+    end
   end
```

Reply:
Ah, this is great! Thanks –– updated the factories to do this